File: lista.cpp
#include <iostream>

class Nodo {
public:
  int item;
  Nodo *siguiente;

  Nodo(int valor) : item(valor), siguiente(nullptr){};
};

class Cola {
private:
  Nodo *cabeza;
  int cantidad;

public:
  Cola() : cabeza(nullptr), cantidad(0) {}

  ~Cola() {
    while (!vacia()) {
      suprimir();
    }
  }

  void insertar(int item) {
    Nodo *nuevoNodo = new Nodo(item);
    nuevoNodo->siguiente = cabeza;
    cabeza = nuevoNodo;
    cantidad++;
  }

  void suprimir() {
    if (vacia()) {
      std::cout << "Pila vacia." << std::endl;
    } else {
      Nodo *nodoAEliminar = cabeza;
      cabeza = cabeza->siguiente;
      delete nodoAEliminar;
      cantidad--;
    }
  }

  void recorrer() {
    Nodo *actual = cabeza;
    while (actual != nullptr) {
      std::cout << actual->item << " ";
      actual = actual->siguiente;
    }
    std::cout << std::endl;
  }

  bool vacia() { return cantidad == 0; }
};

void conversionBinaria(int decimal) {
  Cola pila;

  int valorActual = decimal;
  while (valorActual >= 2) {
    pila.insertar(valorActual % 2);
    valorActual /= 2;
  }
  pila.insertar(valorActual % 2);

  while (!pila.vacia()) {
  }
  pila.recorrer();
}

int main() {
  conversionBinaria(23519);
  return 0;
}

File: python/16-encadenamiento/lista.py
class Nodo:
    def __init__(self, item, siguiente=None) -> None:
        self.__item = item
        self.__siguiente = siguiente

    def get_siguiente(self):
        return self.__siguiente

    def get_item(self):
        return self.__item

    def set_siguiente(self, siguiente):
        self.__siguiente = siguiente

    def set_item(self, item):
        self.__item = item


class Lista:
    __cantidad: int
    __cabeza: Nodo | None
    __lista: Nodo | None

    def __init__(self) -> None:
        self.__cantidad = 0
        self.__cabeza = None
        self.__lista = None

    def insertar(self, item: object) -> None:
        nuevo_nodo = Nodo(item)
        actual = self.__cabeza
        anterior = None

        while actual is not None and actual.get_item() <= item:
            anterior = actual
            actual = actual.get_siguiente()

        if anterior is None:
            nuevo_nodo.set_siguiente(self.__cabeza)  # 👈 Enlaza el resto de la lista
            self.__cabeza = nuevo_nodo
        else:
            nuevo_nodo.set_siguiente(actual)
            anterior.set_siguiente(nuevo_nodo)
        self.__cantidad += 1

    # def insertar(self, item: object) -> None:
    #     nuevo_nodo = Nodo(item)
    #     actual = self.__cabeza
    #     anterior = None
    #
    #     while actual is not None and actual.get_item() <= item:
    #         anterior = actual
    #         actual = actual.get_siguiente()
    #
    #     if anterior is None:
    #         self.__cabeza = nuevo_nodo
    #     else:
    #         nuevo_nodo.set_siguiente(actual)
    #         anterior.set_siguiente(nuevo_nodo)
    #     self.__cantidad += 1

    def buscar(self, item) -> int:
        actual = self.__cabeza
        i = 0

        while actual is not None and actual.get_item() != item:
            actual = actual.get_siguiente()
            i += 1

        return i if actual is not None else -1

    def mostrar(self):
        actual = self.__cabeza
        while actual is not None:
            print(actual.get_item(), end=" ")
            actual = actual.get_siguiente()
        print()
